### PR TITLE
add is_sorted, is_sorted_by, is_sorted_by_key

### PR DESCRIPTION
**Pull request by gnzlbg**

### CODE REVIEW

**Comment by gnzlbg** [comment]
@bluss looks like rand made a breaking change but did not bump the major semver version :(

File: src/lib.rs [hunk 1]
```diff
@@ -953,6 +953,77 @@ pub trait Itertools : Iterator {
...
+    fn is_sorted_by<F>(mut self, mut compare: F) -> bool
+        where Self: Sized,
+              Self::Item: Ord,
+              F: FnMut(&Self::Item, &Self::Item) -> Ordering, 
+    {
+        let first = self.next();
+        if let Some(mut first) = first {
+            while let Some(second) = self.next() {
+                if compare(&first, &second) == Ordering::Greater {
+                    return false;
```
Comment (scottmcm) [inline src/lib.rs]: For a while-let with an early return, should this be `&mut self` instead, like `all` and `any`?

(nit: can it be written [in terms of `all`](https://medium.com/@veedrac/rust-is-slow-and-i-am-the-cure-32facc0fdcb#9a37), until `try_fold` is stable?)

Comment (gnzlbg) [reply src/lib.rs]: I'll give these a try!

Comment (gnzlbg) [reply src/lib.rs]: @scottmcm these should be fixed now

**Comment by gnzlbg** [comment]
@scottmcm probably the solution to fixing the travis bots is to raise the minimum version of stable supported, or to disable tests that don't work on the stable version used. Bumping the minimum version of stable supported will require a major semver bump though (0.7 to 0.8).

**Comment by bluss** [comment]
sad to see rand breaking that

**Comment by bluss** [comment]
Hi, it looks good. Please place these new methods in the regular method section. For example after the sorted methods.

**Comment by gnzlbg** [comment]
I've moved the methods. Is that place ok? I did not realize the methods were separated into sections :/

**Comment by bluss** [comment]
It's in the Itertools trait doc, but it's impossible to see that doc due to the big ugly code block that starts a trait documentation page.

**Comment by bluss** [comment]
> The trait Itertools: extra iterator adaptors and methods for iterators.

> This trait defines a number of methods. They are divided into two groups:

>    Adaptors take an iterator and parameter as input, and return a new iterator value. These are listed first in the trait. An example of an adaptor is .interleave()

>    Regular methods are those that don't return iterators and instead return a regular value of some other kind. .next_tuple() is an example and the first regular method in the list.

**Comment by bluss** [comment]
Thanks. We need to see how the discussion for std is going. As witnessed with the flatten problems, it would suck to introduce these here only to have them in std shortly afterwards.

**Comment by gnzlbg** [comment]
No worries, I have my own fork of itertools in the meantime and I can
migrate to nightly if these are incuded in std in the mean time. I just
wanted to upstream these here to show exactly which implementation I was
proposing in the Rust-Lang issue.

There were some questions about making these free functions instead but I
still do think that making these Iterator methods is the most flexible
approach.

**Comment by bluss** [comment]
Superseded by work in std, rust-lang/rust/pull/55045

**Pull request closed without merging**